File: Source/Pegasus/Pegasus/Platform/Graphics/Direct3D11/Texture.D3D11.cs
namespace Pegasus.Platform.Graphics.Direct3D11
{
	using System;
	using Bindings;
	using Interface;

	/// <summary>
	///     Represents an Direct3D11-based texture stored in GPU memory.
	/// </summary>
	internal class TextureD3D11 : GraphicsObjectD3D11, ITexture
	{
		/// <summary>
		///     The underlying Direct3D11 shader resource view.
		/// </summary>
		private D3D11ShaderResourceView _view;

		/// <summary>
		///     Gets the underlying Direct3D11 texture.
		/// </summary>
		internal D3D11Resource Texture;

		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		/// <param name="graphicsDevice">The graphics device this instance belongs to.</param>
		/// <param name="description">The description of the texture.</param>
		/// <param name="surfaces">The surface data of the texture.</param>
		public TextureD3D11(GraphicsDeviceD3D11 graphicsDevice, ref TextureDescription description, Surface[] surfaces)
			: base(graphicsDevice)
		{
			switch (description.Type)
			{
				case TextureType.Texture2D:
					CreateTexture2D(ref description, surfaces);
					break;
				case TextureType.CubeMap:
					CreateCubeMap(ref description, surfaces);
					break;
				default:
					throw new InvalidOperationException("Unsupported texture type.");
			}
		}

		/// <summary>
		///     Binds the texture to the given slot.
		/// </summary>
		/// <param name="slot">The slot the texture should be bound to.</param>
		public unsafe void Bind(int slot)
		{
			var view = _view;
			Context.VSSetShaderResources(slot, 1, &view);
			Context.PSSetShaderResources(slot, 1, &view);
		}

		/// <summary>
		///     Unbinds the texture from the given slot.
		/// </summary>
		/// <param name="slot">The slot the texture should be unbound from.</param>
		public unsafe void Unbind(int slot)
		{
			Context.VSSetShaderResources(slot, 0, null);
			Context.PSSetShaderResources(slot, 0, null);
		}

		/// <summary>
		///     Generates the mipmaps for this texture.
		/// </summary>
		public void GenerateMipmaps()
		{
			Context.GenerateMips(_view);
		}

		/// <summary>
		///     Sets the debug name of the texture.
		/// </summary>
		/// <param name="name">The debug name of the texture.</param>
		public void SetName(string name)
		{
			Texture.SetDebugName(name);

			if (_view.IsInitialized)
				_view.SetDebugName(name);
		}

		/// <summary>
		///     Initializes a 2D texture.
		/// </summary>
		/// <param name="description">The description of the texture.</param>
		/// <param name="surfaces">The surface data of the texture.</param>
		private unsafe void CreateTexture2D(ref TextureDescription description, Surface[] surfaces)
		{
			D3D11Texture2DDescription desc;
			InitializeDesc2D(ref description, out desc);

			var data = stackalloc D3D11SubResourceData[GraphicsDevice.MaxSurfaceCount];
			InitializeData(data, ref description, surfaces);
			Device.CreateTexture2D(ref desc, surfaces == null ? null : data, out Texture).CheckSuccess("Failed to create 2D texture.");

			if ((description.Flags & TextureFlags.DepthStencil) == TextureFlags.DepthStencil)
				return;

			var viewDesc = new D3D11ShaderResourceViewDescription
			{
				Format = desc.Format,
				Dimension = D3D11ShaderResourceViewDimension.Texture2D,
				Texture2D =
				{
					MipLevels = -1,
					MostDetailedMip = 0
				}
			};

			Device.CreateShaderResourceView(Texture, ref viewDesc, out _view).CheckSuccess("Failed to create shader resource view.");
		}

		/// <summary>
		///     Initializes a cube map.
		/// </summary>
		/// <param name="description">The description of the texture.</param>
		/// <param name="surfaces">The surface data of the texture.</param>
		private unsafe void CreateCubeMap(ref TextureDescription description, Surface[] surfaces)
		{
			D3D11Texture2DDescription desc;
			InitializeDesc2D(ref description, out desc);
			desc.ArraySize *= 6;
			desc.OptionFlags |= D3D11ResourceOptionFlags.TextureCube;

			var data = stackalloc D3D11SubResourceData[GraphicsDevice.MaxSurfaceCount];
			InitializeData(data, ref description, surfaces);
			Device.CreateTexture2D(ref desc, surfaces == null ? null : data, out Texture).CheckSuccess("Failed to create 2D texture.");

			if ((description.Flags & TextureFlags.DepthStencil) == TextureFlags.DepthStencil)
				return;

			var viewDesc = new D3D11ShaderResourceViewDescription
			{
				Format = desc.Format,
				Dimension = D3D11ShaderResourceViewDimension.TextureCube,
				Texture2D =
				{
					MipLevels = -1,
					MostDetailedMip = 0
				}
			};

			Device.CreateShaderResourceView(Texture, ref viewDesc, out _view).CheckSuccess("Failed to create shader resource view.");
		}

		/// <summary>
		///     Initializes a 2D texture description.
		/// </summary>
		/// <param name="description">The Pegasus texture description.</param>
		/// <param name="desc">The initialized Direct3D11 description.</param>
		private static void InitializeDesc2D(ref TextureDescription description, out D3D11Texture2DDescription desc)
		{
			desc.Width = (int)description.Width;
			desc.Height = (int)description.Height;
			desc.MipLevels = (int)description.Mipmaps;
			desc.ArraySize = 1;
			desc.SampleDescription.Count = 1;
			desc.SampleDescription.Quality = 0;
			desc.Usage = D3D11ResourceUsage.Default;
			desc.BindFlags = D3D11BindFlags.ShaderResource;
			desc.CpuAccessFlags = D3D11CpuAccessFlags.None;
			desc.OptionFlags = D3D11ResourceOptionFlags.None;
			desc.Format = description.Format.Map();

			if ((description.Flags & TextureFlags.RenderTarget) == TextureFlags.RenderTarget)
				desc.BindFlags |= D3D11BindFlags.RenderTarget;

			if ((description.Flags & TextureFlags.DepthStencil) == TextureFlags.DepthStencil)
				desc.BindFlags = D3D11BindFlags.DepthStencil;

			if ((description.Flags & TextureFlags.GenerateMipmaps) == TextureFlags.GenerateMipmaps)
			{
				desc.OptionFlags |= D3D11ResourceOptionFlags.GenerateMipMaps;
				desc.BindFlags |= D3D11BindFlags.RenderTarget;
				desc.MipLevels = 0;
			}
		}

		/// <summary>
		///     Initializes the texture data.
		/// </summary>
		/// <param name="data">The texture data that should be initialized.</param>
		/// <param name="description">The description of the texture.</param>
		/// <param name="surfaces">The surfaces of the texture.</param>
		private static unsafe void InitializeData(D3D11SubResourceData* data, ref TextureDescription description, Surface[] surfaces)
		{
			if (surfaces == null)
				return;

			var mipmaps = description.Mipmaps;
			if ((description.Flags & TextureFlags.GenerateMipmaps) == TextureFlags.GenerateMipmaps)
				mipmaps = 0;

			if (description.Type == TextureType.CubeMap)
			{
				int* faces = stackalloc int[6];
				faces[0] = 5;
				faces[1] = 1;
				faces[2] = 4;
				faces[3] = 0;
				faces[4] = 3;
				faces[5] = 2;

				for (var i = 0; i < 6; ++i)
				{
					for (var j = 0; j < mipmaps; ++j)
					{
						var dataIndex = faces[i] * mipmaps + j;
						var surfaceIndex = i * mipmaps + j;
						data[dataIndex].Data = new IntPtr(surfaces[surfaceIndex].Data);
						data[dataIndex].RowPitch = (int)surfaces[surfaceIndex].Stride;
						data[dataIndex].SlicePitch = (int)surfaces[surfaceIndex].SizeInBytes;
					}
				}
			}
			else
			{
				for (var i = 0; i < mipmaps; ++i)
				{
					data[i].Data = new IntPtr(surfaces[i].Data);
					data[i].RowPitch = (int)surfaces[i].Stride;
					data[i].SlicePitch = (int)surfaces[i].SizeInBytes;
				}
			}
		}

		/// <summary>
		///     Disposes the object, releasing all managed and unmanaged resources.
		/// </summary>
		protected override void OnDisposing()
		{
			Texture.Release();
			_view.Release();
		}
	}
}

File: Source/Pegasus/Platform/Graphics/Direct3D11/TextureD3D11.cpp
#include "Prelude.hpp"

using namespace Direct3D11;

Texture* GraphicsDevice::InitializeTexture(TextureDescription* desc, Surface* surfaces)
{
	auto texture = PG_NEW(Texture);

	switch (desc->Type)
	{
	case TextureType::Texture1D:
		CreateTexture1D(texture, desc, surfaces);
		break;
	case TextureType::Texture2D:
		CreateTexture2D(texture, desc, surfaces);
		break;
	case TextureType::CubeMap:
		CreateCubeMap(texture, desc, surfaces);
		break;
	case TextureType::Texture3D:
		CreateTexture3D(texture, desc, surfaces);
		break;
	default:
		PG_NO_SWITCH_DEFAULT;
	}

	return texture;
}

void GraphicsDevice::FreeTexture(Texture* texture)
{
	PG_DELETE(texture);
}

void GraphicsDevice::BindTexture(Texture* texture, int32 slot)
{
	_context->VSSetShaderResources(safe_static_cast<UINT>(slot), 1, texture->View.GetAddressOf());
	_context->PSSetShaderResources(safe_static_cast<UINT>(slot), 1, texture->View.GetAddressOf());
}

void GraphicsDevice::UnbindTexture(Texture* texture, int32 slot)
{
	PG_UNUSED(texture);

	ID3D11ShaderResourceView* resourceView[] = { nullptr };
	_context->VSSetShaderResources(safe_static_cast<UINT>(slot), 1, resourceView);
	_context->PSSetShaderResources(safe_static_cast<UINT>(slot), 1, resourceView);
}

void GraphicsDevice::GenerateMipmaps(Texture* texture)
{
	_context->GenerateMips(texture->View.Get());
}

void GraphicsDevice::SetTextureName(Texture* texture, const char* name)
{
	SetName(texture->Obj, name);
	SetName(texture->View, name);
}

void GraphicsDevice::CreateTexture1D(Texture* texture, const TextureDescription* desc, const Surface* surfaces)
{
	PG_UNUSED(texture);
	PG_UNUSED(desc);
	PG_UNUSED(surfaces);

	PG_DIE("Not implemented.");
}

void GraphicsDevice::CreateTexture2D(Texture* texture, const TextureDescription* desc, const Surface* surfaces)
{
	D3D11_SUBRESOURCE_DATA data[Graphics::MaxSurfaceCount];
	InitializeResourceData(data, desc, surfaces);

	D3D11_TEXTURE2D_DESC d3d11Desc;
	InitializeDesc2D(desc, d3d11Desc);

	ID3D11Texture2D* tex2D;
	CheckResult(_device->CreateTexture2D(&d3d11Desc, surfaces == nullptr ? nullptr : data, &tex2D), "Failed to create texture.");
	texture->Obj.Attach(tex2D);

	if (Enum::HasFlag(desc->Flags, TextureFlags::DepthStencil))
		return;

	D3D11_SHADER_RESOURCE_VIEW_DESC viewDesc;
	Memory::Set(&viewDesc, 0);
	viewDesc.Format = d3d11Desc.Format;
	viewDesc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURE2D;
	viewDesc.Texture2D.MipLevels = static_cast<UINT>(-1);
	viewDesc.Texture2D.MostDetailedMip = 0;

	CheckResult(
		_device->CreateShaderResourceView(texture->Obj.Get(), &viewDesc, &texture->View),
		"Failed to create shader resource view for texture.");
}

void GraphicsDevice::CreateTexture3D(Texture* texture, const TextureDescription* desc, const Surface* surfaces)
{
	PG_UNUSED(texture);
	PG_UNUSED(desc);
	PG_UNUSED(surfaces);

	PG_DIE("Not implemented.");
}

void GraphicsDevice::CreateCubeMap(Texture* texture, const TextureDescription* desc, const Surface* surfaces)
{
	D3D11_SUBRESOURCE_DATA data[Graphics::MaxSurfaceCount];
	InitializeResourceData(data, desc, surfaces);

	D3D11_TEXTURE2D_DESC d3d11Desc;
	InitializeDesc2D(desc, d3d11Desc);
	d3d11Desc.ArraySize *= 6;
	d3d11Desc.MiscFlags |= D3D11_RESOURCE_MISC_TEXTURECUBE;

	ID3D11Texture2D* cubeMap;
	CheckResult(_device->CreateTexture2D(&d3d11Desc, surfaces == nullptr ? nullptr : data, &cubeMap), "Failed to create cube map.");
	texture->Obj.Attach(cubeMap);

	if (Enum::HasFlag(desc->Flags, TextureFlags::DepthStencil))
		return;

	D3D11_SHADER_RESOURCE_VIEW_DESC viewDesc;
	Memory::Set(&viewDesc, 0);
	viewDesc.Format = d3d11Desc.Format;
	viewDesc.ViewDimension = D3D11_SRV_DIMENSION_TEXTURECUBE;
	viewDesc.TextureCube.MipLevels = static_cast<UINT>(-1);
	viewDesc.TextureCube.MostDetailedMip = 0;

	CheckResult(
		_device->CreateShaderResourceView(texture->Obj.Get(), &viewDesc, &texture->View),
		"Failed to create shader resource view for cube map.");
}

void GraphicsDevice::InitializeDesc2D(const TextureDescription* desc, D3D11_TEXTURE2D_DESC& d3d11Desc)
{
	Memory::Set(&d3d11Desc, 0);
	d3d11Desc.Width = safe_static_cast<UINT>(desc->Width);
	d3d11Desc.Height = safe_static_cast<UINT>(desc->Height);
	d3d11Desc.MipLevels = safe_static_cast<UINT>(desc->Mipmaps);
	d3d11Desc.ArraySize = 1;
	d3d11Desc.SampleDesc.Count = 1;
	d3d11Desc.SampleDesc.Quality = 0;
	d3d11Desc.Usage = D3D11_USAGE_DEFAULT;
	d3d11Desc.BindFlags = D3D11_BIND_SHADER_RESOURCE;
	d3d11Desc.CPUAccessFlags = 0;
	d3d11Desc.MiscFlags = 0;
	d3d11Desc.Format = Map(desc->Format);

	if (Enum::HasFlag(desc->Flags, TextureFlags::RenderTarget))
		d3d11Desc.BindFlags |= D3D11_BIND_RENDER_TARGET;

	if (Enum::HasFlag(desc->Flags, TextureFlags::DepthStencil))
		d3d11Desc.BindFlags = D3D11_BIND_DEPTH_STENCIL;

	if (Enum::HasFlag(desc->Flags, TextureFlags::GenerateMipmaps))
	{
		d3d11Desc.MiscFlags |= D3D11_RESOURCE_MISC_GENERATE_MIPS;
		d3d11Desc.BindFlags |= D3D11_BIND_RENDER_TARGET;
		d3d11Desc.MipLevels = 0;
	}
}

void GraphicsDevice::InitializeResourceData(D3D11_SUBRESOURCE_DATA* resourceData, const TextureDescription* desc, const Surface* surfaces)
{
	if (surfaces == nullptr)
		return;

	auto mipmaps = desc->Mipmaps;
	if (Enum::HasFlag(desc->Flags, TextureFlags::GenerateMipmaps))
		mipmaps = 0;

	if (desc->Type == TextureType::CubeMap)
	{
		int32 faces[] = { 5, 1, 4, 0, 3, 2 };

		for (auto i = 0; i < 6; ++i)
		{
			for (auto j = 0; j < mipmaps; ++j)
			{
				auto d3dIndex = faces[i] * mipmaps + j;
				auto pgIndex = i * mipmaps + j;
				resourceData[d3dIndex].pSysMem = surfaces[pgIndex].Data;
				resourceData[d3dIndex].SysMemPitch = safe_static_cast<UINT>(surfaces[pgIndex].Stride);
				resourceData[d3dIndex].SysMemSlicePitch = safe_static_cast<UINT>(surfaces[pgIndex].SizeInBytes);
			}
		}
	}
	else
	{
		for (auto i = 0; i < mipmaps; ++i)
		{
			resourceData[i].pSysMem = surfaces[i].Data;
			resourceData[i].SysMemPitch = safe_static_cast<UINT>(surfaces[i].Stride);
			resourceData[i].SysMemSlicePitch = safe_static_cast<UINT>(surfaces[i].SizeInBytes);
		}
	}
}
